lumicks/pylake/kymotracker/detail/gaussian_mle.py

```diff
@@ -0,0 +1,138 @@
+import numpy as np
+from functools import partial
+from scipy.optimize import minimize
+
+
+def poisson_log_likelihood(params, expectation_fcn, photon_count):
+    """Calculates the log likelihood of Poisson distributed values.
+
+    Parameters
+    ----------
+    params : array-like
+        Model parameters to be estimated
+    x : array-like
+        Position data at which the function is to be evaluated.
+    photon_count : array-like
+        Measured photon counts at each position to be fitted.
+    """
+    expectation = expectation_fcn(*params)
+    log_likelihood = photon_count * np.log(expectation) - expectation
+    return np.sum(-log_likelihood)
+
+
+def poisson_log_likelihood_jacobian(params, expectation_fcn, derivatives_fcn, photon_count):
+    """Evaluate the derivatives of the likelihood function w.r.t. each parameter."""
+    derivatives = derivatives_fcn(*params)
+    count_over_expectation = photon_count / expectation_fcn(*params)
+    return [-np.sum(d * ((count_over_expectation) - 1)) for d in derivatives]
+
+
+def _run_mle_optimization(initial_guess, expectation_fcn, derivatives_fcn, photon_count, bounds):
+    """Calculate the maximum likelihood estimate of the model parameters given measured photon count.
+
+    This function is meant to be generalizable for use with either 1D or 2D functions.
+
+    Parameters
+    ----------
+    initial_guess : array-like
+        Initial guesses for the model parameters.
+    expectation_fcn : function
+        expectation function to be evaluated in the likelihood; takes model parameters as arguments.
+    derivatives_fcn : function
+        expectation derivatives function to be evaluaged in the Jacobian; takes model parameters as arguments.
+    photon_count : array-like
+        Measured photon counts at each position to be fitted.
+    bounds : tuple of 2-value tuples
+        Tuple of lower and upper bounds for each parameter.
+    """
+    optim_fcn = partial(
+        poisson_log_likelihood, expectation_fcn=expectation_fcn, photon_count=photon_count
+    )
+    jac_fcn = partial(
+        poisson_log_likelihood_jacobian,
+        expectation_fcn=expectation_fcn,
+        derivatives_fcn=derivatives_fcn,
+        photon_count=photon_count,
+    )
+
+    return minimize(
+        optim_fcn,
+        initial_guess,
+        jac=jac_fcn,
+        method="L-BFGS-B",
+        bounds=bounds,
+    )
+
+
+def normal_pdf_1d(x, center, width):
+    """Evaluates the probablitity density function for a 1D normal distribution
+
+    Parameters
+    ----------
+    x : array-like
+        Position data at which the function is to be evaluated.
+    center : float
+        Distribution center.
+    width : float
+        Distribution width; square root of the variance.
```

Comment:
Personally I prefer `sigma` as a name for this, since it is less ambiguous to me. Width always gets me confused with FWHM. Not a dealbreaker for me though :)

Reply:
I do too, but there have been discussions before about using more descriptive names.

Here I changed it to `sigma` because it is such a well known parameter for the normal distribution, I think it makes sense to use it and avoids confusion with FWHM like you pointed out.